### ISSUE TITLE
Make "d y" call a command without defining a text object

### ISSUE THREAD

**Issue by GaryGlitter** [emacs-evil/evil#1320]
# Issue type

- Question

Hi! I want "d y" to call paredit-kill, how can I do that? I want to preserve the usual behaviour of "y" to evil-yank when not prepended by d, though. This didn't work, the error says "d y" doesn't start w/ a prefix key:

  (define-key evil-normal-state-map (kbd "d y") 'paredit-kill)

Would be glad to get some help on this.

**Comment by TheBB** [comment]
It's not really possible to do this in a generic way with the way evil is currently implemented. The `d` key is mapped to `evil-delete` which then reads another key to determine what to delete.

Evil-surround uses a hack to do something similar. Since operations like `cs` and `ds` aren't implementable as straight up text objects, it maps `s` is in `evil-operator-state` to a command that checks what the current operation is, and then dispatches accordingly, so that `cs` and `ds` can call different commands.

**Comment by GaryGlitter** [comment]
@TheBB hmm maybe I can then defadvice evil-delete somehow, you think? And if not that, am also willing to get a little dirty and hack up evil-delete (but how though?)

**Comment by GaryGlitter** [comment]
@noctuid It works! But what's this `menu-item'? Can't find it in apropos. Also, when I return a command which deletes something, the removed text shows up in the kill ring, but can't be pasted with p (it pastes nothing), unlike when simply calling the command manually.

```
(defvar evil-collection-delete-operators '(evil-delete
                                             evil-cp-delete
                                             evil-sp-delete
                                             lispyville-delete)
    "List of delete operators.")
  
  (evil-define-minor-mode-key 'operator 'evil-mode
    "p" '(menu-item
          ""
          nil
          :filter (lambda (&optional _)
                    (when (memq evil-this-operator
                                evil-collection-delete-operators)
                      #'kill-line))))
```

**Comment by GaryGlitter** [comment]
@noctuid

Figured menu-item is [https://www.gnu.org/software/emacs/manual/html_node/elisp/Extended-Menu-Items.html](this).

Maan do I feel dumb. Ofc evil-delete is _still_ running when the command is called and so deletes to where the cursor has traveled after the command is done. Calling (keyboard-quit) seems to solve the issue:

```
(evil-define-minor-mode-key 'operator 'evil-mode
  "y" '(menu-item
        ""
        nil
        :filter (lambda (&optional _)
                  (when (memq evil-this-operator
                              evil-collection-delete-operators)
                    (lambda () (interactive) (paredit-kill) (keyboard-quit))))))
```

`keyboard-quit` looks pretty weird and I wonder if there's a more elegant way to cancel evil-delete, but the key remapping works and so I close the issue.

> general-key-dispatch

Out of curiosity, what's the correct way to use it? I am so far at this point: `(general-key-dispatch nil (kbd "dy") 'paredit-kill)`

Since it returns a command, I tried doing command-execute on it and sure enough it allows accepting dy right afterwards, but it's not permanent.
